Revamp of manual grading instance question page

jonatanschroeder opened the pull request:
Rewriting in TS, with type validation and new processes.

jonatanschroeder commented:
There are so many changes here that I'm avoiding the two-stage (change then type) for simplicity. I know this loses history, but I don't think history would be very beneficial here for most cases, given minor changes everywhere.

jonatanschroeder commented:
Adding here to persist the discussion, but I asked the same question on Slack: the manual grading interface allows instructors to use either points or percentage to for manual grading. The functionality of percentage is trivial when only manual points are assigned to a question (i.e., no auto-points), but when auto-points are involved:

* If there are both manual points and auto-points, then the percentage assigned in manual grading is based on that component. For example, if there are 5 manual points and 10 auto points, then assigning 40% in manual score will trigger 2 manual points, and 70% auto score will trigger 7 auto-points, for a total of 9 points (which is shown as 9/15=60% as the total score in the interface). I think this makes sense, and can continue as is. This is the same in the manual grading interface and in upload when using columns like `manual_score_perc` or `auto_score_perc`, while setting `score_perc` changes manual points such that the total ends up equivalent to the requested percentage, and is internally consistent, afaict.
* If there are only auto-points, then the percentage in manual score is currently inconsistent. Setting the manual score to any percentage has no bearing on points, since the percentage is applied to manual points, which are zero. This is done both [client-side (when updating the viewed score)](https://github.com/PrairieLearn/PrairieLearn/blob/486f8c551f3d04a62a6031cc744f4b39021154f1/apps/prairielearn/assets/scripts/instructorAssessmentManualGradingInstanceQuestion.js#L363) and [server-side](https://github.com/PrairieLearn/PrairieLearn/blob/486f8c551f3d04a62a6031cc744f4b39021154f1/apps/prairielearn/src/lib/manualGrading.ts#L541), and also affects score upload. However, if the manual points had already been set beforehand (e.g., by using points), then the [percentage shown in manual score when the page is loaded](https://github.com/PrairieLearn/PrairieLearn/blob/486f8c551f3d04a62a6031cc744f4b39021154f1/apps/prairielearn/assets/scripts/instructorAssessmentManualGradingInstanceQuestion.js#L368) is based on total points instead of manual points.

Question is: which behaviour should remain in this case? My suggestion would be to disable percentage-based grading in the manual grading UI if there are no manual points, and only allow points-based grading in that scenario. This should remove any confusion or misconception on the meaning of the percentage in this case.

jonatanschroeder commented on apps/prairielearn/assets/scripts/instructorAssessmentManualGradingInstanceQuestion.js:
```diff
@@ -89,10 +89,13 @@ function resetInstructorGradingPanel() {
+      form.querySelectorAll('.js-auto-score-value-input').forEach((inputDiv) => {
+        inputDiv.classList.remove('d-none');
+        inputDiv.style.display = '';
+        inputDiv.querySelectorAll('input').forEach((input) => {
+          input.name = input.dataset.hiddenInputName;
```
This is a slight changed behaviour to account for the Zod validation. In essence, the original code in the server would use `req.body.score_auto_points || null` on the understanding that, if the pencil was never clicked, the `score_auto_points` value would not be sent, which was a mistake, since the input exists but is hidden. This never caused an issue before because the `score_auto_points` is populated with the original auto-points, meaning that manual grading would be set with the auto-points that already existed, so manual grading would typically update it with its current value instead of keeping it as it is (in essence, setting it to the value it already had).

The problem happens when trying to change the auto points from non-zero to zero. Before, this would work because the input would provide `'0'`, which is truthy, but after Zod validation this is now parsed to `0`, which is falsy.

The solution here is that the auto-points input is initially created without a name, but with a `data-hidden-input-name` instead, and when it is enabled the name is set as needed. This will cause the original intended functionality, which is that if pencil is not selected, `score_auto_points` is never sent as a body parameter, and can then be handled as null instead of falsy.

jonatanschroeder commented on apps/prairielearn/assets/scripts/instructorAssessmentManualGradingInstanceQuestion.js:
```diff
@@ -353,40 +352,42 @@ function updatePointsView(sourceInput) {
     const auto_points =
       roundPoints(
-        sourceInput?.name === 'score_auto_percent'
+        sourceInput?.classList?.contains('js-auto-score-value-input-percentage')
           ? (sourceInput?.value * max_auto_points) / 100
-          : form.querySelector('[name=score_auto_points]')?.value,
+          : form.querySelector('.js-auto-score-value-input-points')?.value,
       ) || 0;
```
Because the name is no longer always there (see other comment), this query now uses a class instead (which already existed anyway).

nwalters512 commented:
> There are so many changes here that I'm avoiding the two-stage (change then type) for simplicity.

The other reason to prefer a two-stage change is that you get generally-useful diffs for review. Candidly, this PR is _really_ difficult to review in its current form. I understand that we won't have any useful diffs for the EJS->TS conversions, but for the Express router handlers, everything just shows up as brand new code. I'm not sure how to start reviewing this.

If you'd really prefer to avoid splitting the work, I'll still do my best to review it.

jonatanschroeder commented:
> > There are so many changes here that I'm avoiding the two-stage (change then type) for simplicity.
> 
> The other reason to prefer a two-stage change is that you get generally-useful diffs for review. Candidly, this PR is _really_ difficult to review in its current form. I understand that we won't have any useful diffs for the EJS->TS conversions, but for the Express router handlers, everything just shows up as brand new code. I'm not sure how to start reviewing this.
> 
> If you'd really prefer to avoid splitting the work, I'll still do my best to review it.

I renamed the router handler file back to its original name and added JSDoc comments. Not much I can do with the EJS->TS in that direction.

nwalters512 reviewed (approved):
I skimmed through these changes, they seem reasonable. I gave myself additional confidence in the conversion by diffing the Prettier-ified HTML of the manual grading page before and after this change and seeing that the generated HTML is basically identical, with just a few exceptions that looked fine.
